src/olmo_eval/metrics.py

```diff
@@ -121,37 +169,57 @@ def compute(self) -> torch.Tensor:
 
             skip_document = False
             for cont_id in loglikelihood_dict[doc_id]:
                 try:
                     loglikelihoods[cont_id] = loglikelihood_dict[doc_id][cont_id]
+                    celosses[cont_id] = celoss_dict[doc_id][cont_id]
+                    bpbs[cont_id] = bpb_dict[doc_id][cont_id]
                 except IndexError:
                     # We didn't process all of the continuations, so skip this document.
                     skip_document = True
                     break
 
             if skip_document:
                 continue
-            if self.metric_type in ["ce_loss", "bpb"]:
-                correct.append(loglikelihoods[0])  # Only one answer is scored
-            else:
-                correct.append(
-                    1.0 if torch.argmax(loglikelihoods).item() == label_dict[doc_id] else 0.0
-                )
-
-            if self.metric_type == "f1":
+            if self.metric_type == "ce_loss":
+                celoss.append(celosses[0])  # Only one answer is scored
+            elif self.metric_type == "bpb":
+                bpb.append(bpbs[0])  # Only one answer is scored
+            elif self.metric_type == "f1":
                 assert preds is not None
                 assert labels is not None
                 preds.append(torch.argmax(loglikelihoods).item())
                 labels.append(label_dict[doc_id])
+            else:
+                correct.append(
+                    1.0 if torch.argmax(loglikelihoods).item() == label_dict[doc_id] else 0.0
+                )
+                celoss.append(celosses[label_dict[doc_id]].item())
+                bpb.append(bpbs[label_dict[doc_id]].item())
+                soft_score.append(torch.softmax(loglikelihoods, dim=0)[label_dict[doc_id]].item())
+                soft_log_score.append(
+                    torch.log_softmax(loglikelihoods, dim=0)[label_dict[doc_id]].item()
+                )
 
         if self.metric_type == "f1":
             assert preds is not None
             assert labels is not None
             # for NLI tasks, continuations are yes, no, neither, so idx=0 assigned to pos label
             score = f1_score(labels, preds, pos_label=0)
+            return {"f1": torch.tensor(score)}
+        elif self.metric_type == "ce_loss":
+            return {"ce_loss": torch.tensor(sum(celoss) / len(celoss))}
+        elif self.metric_type == "bpb":
+            return {"bpb": torch.tensor(sum(bpb) / len(bpb))}
         else:
-            score = sum(correct) / len(correct)
-
-        return torch.tensor(score)
+            return {
+                self.metric_type: torch.tensor(sum(correct) / len(correct)),
+                "ce_loss": torch.tensor(sum(celoss) / len(celoss)),
+                "bpb": torch.tensor(sum(bpb) / len(bpb)),
+                "soft": torch.tensor(sum(soft_score) / len(soft_score)),
+                "soft_log": torch.tensor(sum(soft_log_score) / len(soft_log_score)),
+            }
```

Comment:
This new, richer metric output (along with the correlated changes in olmo-core's [evaluator_callback.py](https://github.com/allenai/OLMo-core/pull/122/files#diff-b8f5a9e2be0f3b2168530bdfc2a93ccde1b8e29d98b069862439fffd58d1039dL218)) will change the recorded metrics for "acc" and "len_norm" metrics. Do we know that this won't interfere with other parts of the setup? (logging etc)

In general, it's definitely the right thing to do though, to compute these in a single go, rather than as separate versions of the same task (in fact, could also have "acc" and "len_norm" together), other than making it less clear which one is the "official" metric for each task.

Reply:
Ah that makes a lot of sense! I have updated my edits in olmo-core so that it displays the mapped metric type text.